working_with_the_llvm_ast.md

```diff
@@ -0,0 +1,215 @@
+## Working With the LLVM AST and Using it to Make Changes to Clang-Tidy Source
+
+There are situations where a particular clang-tidy check will report a lot of false positives or it will require some
+additions to add more features to it based on the usecase. Solving either problem will require making changes to
+clang-tidy source and require both understanding the AST and working with the AST to make code changes as all the
+matchers in clang-tidy source are reliant on the AST and written using the AST. The following sections go into resources
+and examples that will be helpful in gaining a better understanding of the LLVM AST and writing code using it. 
+
+### Understanding the AST
+
+The following list of resources goes into the basics and details of understanding the LLVM AST - 
+
+1. LLVM AST basics, traversing the AST, important terms and examples. The link also contains a video on the AST and is
+   useful to watch - [Understanding the Clang AST](https://jonasdevlieghere.com/understanding-the-clang-ast/).
+
+2. Official LLVM Tutorial on the AST. The tutorial contains links to important concepts like ASTContext, AST Nodes and
+   are a must read - [Official LLVM AST Tutorial](https://clang.llvm.org/docs/IntroductionToTheClangAST.html).
+
+3. LLVM Doxygen Documentation for the AST. This serves as a reference for all the LLVM AST specific classes and their
+   methods. This is a very, very important reference when writing or reading any AST specific code and has details on
+   methods that might be useful in writing code that uses the AST or for writing your own AST Matchers - [LLVM AST
+   Classes Documentation](https://clang.llvm.org/doxygen/namespaceclang.html).
+
+4. LLVM AST Matchers Reference. Very important reference that documents all the available matchers, their syntax and the
+   correct parameters that they accept - [LLVM AST Matchers
+   Reference](https://clang.llvm.org/docs/LibASTMatchersReference.html).
+
+5. Writing a Custom Clang-Tidy Check. Although this section deals with resources for understanding the AST, it is also
+   geared towards using that knowledge to write custom clang-tidy checks. This resource is helpful to understand the
+   structure of a clang-tidy check and how to write custom checkers, or use that knowledge to modify existing checkers -
+   [Writing a Custom Clang-Tidy Check](http://blog.audio-tk.com/2018/03/20/writing-custom-checks-for-clang-tidy).
+
+### How to Inspect/Verify New AST Matchers
+
+You have just written an AST matcher to match something in your code but want to verify quickly if the matcher indeed
+matches what it is supposed to? One way to do that is to build clang-tidy and debug the matcher using a regression test
+to see whether the matcher returns the correct thing. This works but for that one needs to build the source, run the
```

Comment:
I'd avoid using one as a reference to a person. Documentation works best when natural language is used, as if you're speaking directly to the reader. Consider using `you'd`. Same goes for the rest of the doc.